pytorch_toolkit/formula_recognition/tools/test.py

```diff
@@ -0,0 +1,297 @@
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from tools.utils.common import (DECODER_INPUTS, DECODER_OUTPUTS,
+                                ENCODER_INPUTS, ENCODER_OUTPUTS, read_net)
+from tools.utils.evaluation_utils import Im2latexRenderBasedMetric
+from tools.utils.get_config import get_config
+
+spaces = [r'\,', r'\>', r'\;', r'\:', r'\quad', r'\qquad', '~']
+
+
+def ends_with_space(string):
+    """If string end with one of the latex spaces (given the above),
+    returns True and index of this space, else False and None
+
+    Args:
+        string (str): input string with possible spaces
+
+    Returns:
+        Tuple(bool, int) string ends with space, index of the space
+    """
+    for idx, space in enumerate(spaces):
+        if string.endswith(space):
+            return True, idx
+    return False, None
+
+
+def get_onnx_inputs(model):
+    names = []
+    for inp in model.get_inputs():
+        names.append(inp.name)
+    return names
+
+
+def get_onnx_outputs(model):
+    names = []
+    for out in model.get_outputs():
+        names.append(out.name)
+    return names
+
+
+def postprocess_prediction(pred_phrase_str):
+    """Deletes usual space in the end of the string and then checks
+    if string ends with latex space. If yes, deletes latex space.
+    Deletion of spaces is performed because, even though spaces in the end are invisible,
+    they affect on rendering the formula, making it more tight to the left
+
+    Args:
+        pred_phrase_str (str): input string
+
+    Returns:
+        str: postprocessed string
+    """
+    pred_phrase_str = pred_phrase_str.rstrip()
+    ends, idx = ends_with_space(pred_phrase_str)
+    while ends:
+        pred_phrase_str = pred_phrase_str[:len(pred_phrase_str) - len(spaces[idx])]
+        pred_phrase_str = pred_phrase_str.rstrip()
+        ends, idx = ends_with_space(pred_phrase_str)
+    return pred_phrase_str
+
+
+class ExportedModelEvaluator():
+    def __init__(self, config):
+        self.config = config
+        self.load_ir_model()
+        self.load_onnx_model()
+        self.split = config.get("split", "validate")
+        self.val_path = config.get("val_path")
+        self.vocab = read_vocab(config.get('vocab_path'))
+        self.load_dataset()
+        self.read_expected_outputs()
+
+    def load_ir_model(self):
+        ie = IECore()
+        encoder = read_net(self.config.get("res_encoder_name").replace(".onnx", ".xml"), ie)
+        dec_step = read_net(self.config.get("res_decoder_name").replace(".onnx", ".xml"), ie)
+        self.exec_net_encoder = ie.load_network(network=encoder, device_name="CPU")
+        self.exec_net_decoder = ie.load_network(network=dec_step, device_name="CPU")
+
+    def load_onnx_model(self):
+        self.decoder_onnx = onnxruntime.InferenceSession(self.config.get("res_decoder_name"))
+        self.encoder_onnx = onnxruntime.InferenceSession(self.config.get("res_encoder_name"))
+
+    def get_onnx_metric(self):
```

Comment:
Vladimir, as far as I can see there are very similar code in the methods `load_dataset`: it is repeated twice in `ExportedModelEvaluator` and one more time in `Evaluator`

Also there are 3 (three!) very similar methods in this file: `get_onnx_metric` and `get_ir_metric` in the class `ExportedModelEvaluator` and `validate` in `Evaluator`.    
I think they can be merged, if you move into a special inner class the code running the model. You will have three such classes with the same interface for `IR`, `ONNX`, and `Pytorch` cases.
Particularly, the methods `run_ir_model` awill be the method `run` of such new class for `IR`, the combination of `ExportedModelEvaluator.run_encoder` and `ExportedModelEvaluator.run_decoder` will be inside the method `run` for `ONNX`, and the call of the `model(imgs)` will be inside the `run` for `Pytorch` class.    

As the result `Evaluator` and `ExportedModelEvaluator` will be merged.

Also it will allow to merge the methods `load_dataset` into one: it is sufficient to move a similar part of code into a "private" function `_load_dataset` that will have additional parameters for values  `ovino_ir` and `batch_transform` that should be used, after that `load_dataset` will consist of three calls of `_load_dataset` with different parameters.

Also probably several other methods will be merged too.

Reply:
Made `Evaluator` base abstract class, inherited `PyTorchModelEvaluator`, `ONNXModelEvaluator` and `OpenVINOModelEvaluator` from it.